src/consensus/raft/raft.h

```diff
@@ -673,27 +658,31 @@ namespace raft
 
         last_idx = i;
         is_first_entry = false;
-        auto ret = ledger->record_entry(data, size);
+        std::vector<uint8_t> entry;
 
-        if (!ret.second)
+        try
+        {
+          entry = ledger->get_entry(data, size);
+        }
+        catch (const std::logic_error& e)
         {
-          // NB: This will currently never be triggered.
-          // This should only fail if there is malformed data. Truncate
-          // the log and reply false.
+          // This should only fail if there is malformed data.
           LOG_FAIL_FMT(
-            "Recv append entries to {} from {} but the data is malformed",
+            "Recv append entries to {} from {} but the data is malformed: {}",
             local_id,
-            r.from_node);
-
+            r.from_node,
+            e.what());
           last_idx = r.prev_idx;
-          ledger->truncate(r.prev_idx);
           send_append_entries_response(r.from_node, false);
           return;
         }
 
         Term sig_term = 0;
         auto deserialise_success =
-          store->deserialise(ret.first, public_only, &sig_term);
+          store->deserialise(entry, public_only, &sig_term);
+
+        ledger->put_entry(
```

Comment:
Changed the order of things a little bit here: we only write to the ledger after the entry has been deserialised because it is only then that we know that it was a signature transaction.